Index: modules/system/system.admin.inc
===================================================================
RCS file: /cvs/drupal/drupal/modules/system/system.admin.inc,v
retrieving revision 1.327
diff -u -p -r1.327 system.admin.inc
--- modules/system/system.admin.inc	4 Jan 2011 04:02:29 -0000	1.327
+++ modules/system/system.admin.inc	4 Jan 2011 07:10:11 -0000
@@ -783,8 +783,8 @@ function system_modules($form, $form_sta
   $help_arg = module_exists('help') ? drupal_help_arg() : FALSE;
 
   // Used when displaying modules that are required by the install profile.
-  require_once DRUPAL_ROOT . '/includes/install.inc';
-  $distribution_name = check_plain(drupal_install_profile_distribution_name());
+  $profile_info = system_get_info('module', drupal_get_profile()) + array('distribution_name' => 'Drupal');
+  $distribution_name = check_plain($profile_info['distribution_name']);
 
   // Iterate through each of the modules.
   foreach ($visible_files as $filename => $module) {
Index: modules/system/system.install
===================================================================
RCS file: /cvs/drupal/drupal/modules/system/system.install,v
retrieving revision 1.532
diff -u -p -r1.532 system.install
--- modules/system/system.install	2 Jan 2011 17:26:39 -0000	1.532
+++ modules/system/system.install	4 Jan 2011 07:10:12 -0000
@@ -33,7 +33,7 @@ function system_requirements($phase) {
     // is not running the default install profile.
     $profile = drupal_get_profile();
     if ($profile != 'standard') {
-      $info = install_profile_info($profile);
+      $info = system_get_info('module', $profile);
       $requirements['install_profile'] = array(
         'title' => $t('Install profile'),
         'value' => $t('%profile_name (%profile-%version)', array(
