Index: includes/install.core.inc
===================================================================
RCS file: /cvs/drupal/drupal/includes/install.core.inc,v
retrieving revision 1.22
diff -u -p -r1.22 install.core.inc
--- includes/install.core.inc	5 Jun 2010 12:02:33 -0000	1.22
+++ includes/install.core.inc	26 Jun 2010 02:34:43 -0000
@@ -1356,6 +1356,7 @@ function install_profile_modules(&$insta
     'operations' => $operations,
     'title' => st('Installing @drupal', array('@drupal' => drupal_install_profile_distribution_name())),
     'error_message' => st('The installation has encountered an error.'),
+    'finished' => '_install_profile_modules_finished',
   );
   return $batch;
 }
@@ -1518,6 +1519,13 @@ function _install_module_batch($module, 
 }
 
 /**
+ * 'Finished' callback for module installation batch.
+ */
+function _install_profile_modules_finished($success, $results, $operations) {
+  module_clear_major_caches();
+}
+
+/**
  * Checks installation requirements and reports any errors.
  */
 function install_check_requirements($install_state) {
Index: includes/module.inc
===================================================================
RCS file: /cvs/drupal/drupal/includes/module.inc,v
retrieving revision 1.193
diff -u -p -r1.193 module.inc
--- includes/module.inc	24 Jun 2010 17:32:01 -0000	1.193
+++ includes/module.inc	26 Jun 2010 02:34:43 -0000
@@ -173,6 +173,7 @@ function system_list($type) {
  */
 function system_list_reset() {
   drupal_static_reset('system_list');
+  drupal_static_reset('system_region_list');
   drupal_static_reset('list_themes');
   cache_clear_all('bootstrap_modules', 'cache_bootstrap');
   cache_clear_all('system_list', 'cache_bootstrap');
@@ -341,6 +342,10 @@ function module_enable($module_list, $en
 
   $modules_installed = array();
   $modules_enabled = array();
+  foreach (array('system_info_alter', 'node_grants') as $hook) {
+    module_hook_changed($hook, TRUE);
+  }
+
   foreach ($module_list as $module) {
     // Only process modules that are not already enabled.
     $existing = db_query("SELECT status FROM {system} WHERE type = :type AND name = :name", array(
@@ -361,17 +366,8 @@ function module_enable($module_list, $en
         ->condition('type', 'module')
         ->condition('name', $module)
         ->execute();
-      // Refresh the module list to include it.
-      system_list_reset();
-      module_list(TRUE);
-      module_implements('', FALSE, TRUE);
-      _system_update_bootstrap_status();
-      // Update the registry to include it.
-      registry_update();
-      // Refresh the schema to include it.
-      drupal_get_schema(NULL, TRUE);
-      // Clear entity cache.
-      entity_info_cache_clear();
+
+      module_clear_minor_caches($module);
 
       // Now install the module if necessary.
       if (drupal_get_installed_schema_version($module, TRUE) == SCHEMA_UNINSTALLED) {
@@ -394,6 +390,10 @@ function module_enable($module_list, $en
     }
   }
 
+  if (!drupal_installation_attempted()) {
+    module_clear_major_caches();
+  }
+
   // If any modules were newly installed, invoke hook_modules_installed().
   if (!empty($modules_installed)) {
     module_invoke_all('modules_installed', $modules_installed);
@@ -448,13 +448,12 @@ function module_disable($module_list, $d
 
   $invoke_modules = array();
 
+  foreach (array('system_info_alter', 'node_grants') as $hook) {
+    module_hook_changed($hook, TRUE);
+  }
+
   foreach ($module_list as $module) {
     if (module_exists($module)) {
-      // Check if node_access table needs rebuilding.
-      if (!node_access_needs_rebuild() && module_hook($module, 'node_grants')) {
-        node_access_needs_rebuild(TRUE);
-      }
-
       module_load_install($module);
       module_invoke($module, 'disable');
       db_update('system')
@@ -462,29 +461,69 @@ function module_disable($module_list, $d
         ->condition('type', 'module')
         ->condition('name', $module)
         ->execute();
-      $invoke_modules[] = $module;
+
+      module_clear_minor_caches($module);
+
       watchdog('system', '%module module disabled.', array('%module' => $module), WATCHDOG_INFO);
+      $invoke_modules[] = $module;
     }
   }
 
+  module_clear_major_caches();
+
   if (!empty($invoke_modules)) {
-    // Refresh the module list to exclude the disabled modules.
-    system_list_reset();
-    module_list(TRUE);
-    module_implements('', FALSE, TRUE);
-    // Invoke hook_modules_disabled before disabling modules,
-    // so we can still call module hooks to get information.
     module_invoke_all('modules_disabled', $invoke_modules);
-    // Update the registry to remove the newly-disabled module.
-    registry_update();
-    _system_update_bootstrap_status();
   }
+}
 
-  // If there remains no more node_access module, rebuilding will be
-  // straightforward, we can do it right now.
-  if (node_access_needs_rebuild() && count(module_implements('node_grants')) == 0) {
+function module_clear_minor_caches($module) {
+  system_list_reset();
+  module_list(TRUE);
+  module_implements('', FALSE, TRUE);
+  registry_update();
+  drupal_get_schema(NULL, TRUE);
+  entity_info_cache_clear();
+}
+
+function module_clear_major_caches() {
+  // If newly enabled modules implement hook_system_info_alter(), we need to
+  // rebuild the {system} table with new altered info.
+  if (module_hook_changed('system_info_alter')) {
+    system_rebuild_module_data();
+    system_rebuild_theme_data();
+  }
+  if (module_hook_changed('node_grants')) {
     node_access_rebuild();
   }
+  drupal_theme_rebuild();
+  node_types_rebuild();
+  menu_rebuild();
+  cache_clear_all('schema', 'cache');
+  entity_info_cache_clear();
+  drupal_clear_css_cache();
+  drupal_clear_js_cache();
+  _system_update_bootstrap_status();
+
+  // Synchronize to catch any actions that were added or removed.
+  actions_synchronize();
+}
+
+function module_hook_changed($hook, $reset = FALSE) {
+  // We don't use drupal_static() here to avoid the possibility of this list
+  // being externally reset.
+  static $module_list = array();
+
+  $current_module_list = module_implements($hook, TRUE);
+
+  if ($reset) {
+    $module_list[$hook] = $current_module_list;
+    return;
+  }
+
+  if (!isset($module_list[$hook])) {
+    return FALSE;
+  }
+  return $module_list[$hook] != $current_module_list;
 }
 
 /**
Index: modules/simpletest/tests/module_test.module
===================================================================
RCS file: /cvs/drupal/drupal/modules/simpletest/tests/module_test.module,v
retrieving revision 1.6
diff -u -p -r1.6 module_test.module
--- modules/simpletest/tests/module_test.module	22 Apr 2010 18:56:31 -0000	1.6
+++ modules/simpletest/tests/module_test.module	26 Jun 2010 02:34:43 -0000
@@ -36,6 +36,9 @@ function module_test_system_info_alter(&
       $info['dependencies'][] = 'php';
     }
   }
+  if ($file->name == 'seven' && $type == 'theme') {
+    $info['regions']['test_region'] = t('Test region');
+  }
 }
 
 /**
Index: modules/system/system.admin.inc
===================================================================
RCS file: /cvs/drupal/drupal/modules/system/system.admin.inc,v
retrieving revision 1.289
diff -u -p -r1.289 system.admin.inc
--- modules/system/system.admin.inc	21 Jun 2010 02:27:47 -0000	1.289
+++ modules/system/system.admin.inc	26 Jun 2010 02:34:43 -0000
@@ -1229,9 +1229,9 @@ function system_modules_submit($form, &$
   unset($form_state['storage']);
 
   // Installs, enables, and disables modules.
-  module_enable($actions['enable']);
-  module_disable($actions['disable']);
-  module_enable($actions['install']);
+  module_enable($actions['enable'], FALSE);
+  module_disable($actions['disable'], FALSE);
+  module_enable($actions['install'], FALSE);
 
   // Gets module list after install process, displays message if there are changes.
   $post_install_list = module_list(TRUE);
@@ -1239,28 +1239,12 @@ function system_modules_submit($form, &$
     drupal_set_message(t('The configuration options have been saved.'));
   }
 
-  // Clear all caches.
-  registry_rebuild();
-  system_rebuild_theme_data();
-  drupal_theme_rebuild();
-  node_types_rebuild();
-  menu_rebuild();
-  cache_clear_all('schema', 'cache');
-  entity_info_cache_clear();
-  drupal_clear_css_cache();
-  drupal_clear_js_cache();
-
   $form_state['redirect'] = 'admin/modules';
 
   // Notify locale module about module changes, so translations can be
   // imported. This might start a batch, and only return to the redirect
   // path after that.
   module_invoke('locale', 'system_update', $actions['install']);
-
-  // Synchronize to catch any actions that were added or removed.
-  actions_synchronize();
-
-  return;
 }
 
 /**
Index: modules/system/system.module
===================================================================
RCS file: /cvs/drupal/drupal/modules/system/system.module,v
retrieving revision 1.940
diff -u -p -r1.940 system.module
--- modules/system/system.module	24 Jun 2010 17:28:10 -0000	1.940
+++ modules/system/system.module	26 Jun 2010 02:34:44 -0000
@@ -2245,6 +2245,10 @@ function _system_rebuild_module_data() {
     'bootstrap' => 0,
   );
 
+  // Reset the module list to ensure that drupal_alter() runs on a completely
+  // fresh list.
+  module_implements('', FALSE, TRUE);
+
   // Read info files for each module.
   foreach ($modules as $key => $module) {
     // The module system uses the key 'filename' instead of 'uri' so copy the
@@ -2356,6 +2360,10 @@ function _system_rebuild_theme_data() {
     'php' => DRUPAL_MINIMUM_PHP,
   );
 
+  // Reset the module list to ensure that drupal_alter() runs on a completely
+  // fresh list.
+  module_implements('', FALSE, TRUE);
+
   $sub_themes = array();
   // Read info files for each theme
   foreach ($themes as $key => $theme) {
Index: modules/system/system.test
===================================================================
RCS file: /cvs/drupal/drupal/modules/system/system.test,v
retrieving revision 1.127
diff -u -p -r1.127 system.test
--- modules/system/system.test	21 Jun 2010 02:27:47 -0000	1.127
+++ modules/system/system.test	26 Jun 2010 02:34:44 -0000
@@ -1697,6 +1697,74 @@ array_space[a b] = Value';
 }
 
 /**
+ * Tests the effectiveness of hook_system_info_alter().
+ */
+class SystemInfoAlterTestCase extends DrupalWebTestCase {
+  public static function getInfo() {
+    return array(
+      'name' => 'System info alter',
+      'description' => 'Tests the effectiveness of hook_system_info_alter().',
+      'group' => 'System',
+    );
+  }
+
+  /**
+   * Tests that {system}.info is rebuilt after a module that implements
+   * hook_system_info_alter() is enabled. Also tests if core *_list() functions
+   * return freshly altered info.
+   */
+  function testSystemInfoAlter() {
+    // Enable our test module. We do it explicitly instead of using the setUp()
+    // method because setUp() calls system_rebuild_module_data() and could
+    // interfere with our tests.
+    module_enable(array('module_test'), FALSE);
+    $this->assertTrue(module_exists('module_test'), t('Test module is enabled.'));
+
+    $info = $this->_getSystemInfo('seven', 'theme');
+    $this->assertTrue(isset($info['regions']['test_region']), t('Altered theme info was added to {system}.info.'));
+    $seven_regions = system_region_list('seven');
+    $this->assertTrue(isset($seven_regions['test_region']), t('Altered theme info was returned by system_region_list().'));
+    $system_list_themes = system_list('theme');
+    $info = $system_list_themes['seven']->info;
+    $this->assertTrue(isset($info['regions']['test_region']), t('Altered theme info was returned by system_list().'));
+    $list_themes = list_themes();
+    $this->assertTrue(isset($list_themes['seven']->info['regions']['test_region']), t('Altered theme info was returned by list_themes().'));
+
+    // Disable the module and verify that {system}.info is rebuilt without it.
+    // We avoid $disable_dependents on module_disable() because it calls
+    // system_rebuild_module_data() and could interfere with our tests.
+    module_disable(array('module_test'), FALSE);
+    $this->assertFalse(module_exists('module_test'), t('Test module is disabled.'));
+
+    $info = $this->_getSystemInfo('seven', 'theme');
+    $this->assertFalse(isset($info['regions']['test_region']), t('Altered theme info was removed from {system}.info.'));
+    $seven_regions = system_region_list('seven');
+    $this->assertFalse(isset($seven_regions['test_region']), t('Altered theme info was not returned by system_region_list().'));
+    $system_list_themes = system_list('theme');
+    $info = $system_list_themes['seven']->info;
+    $this->assertFalse(isset($info['regions']['test_region']), t('Altered theme info was not returned by system_list().'));
+    $list_themes = list_themes();
+    $this->assertFalse(isset($list_themes['seven']->info['regions']['test_region']), t('Altered theme info was not returned by list_themes().'));
+  }
+
+  /**
+   * Support function. Returns the info array as it is stored in {system}.
+   *
+   * @param $name
+   *   The name of the record in {system}.
+   * @param $type
+   *   The type of record in {system}.
+   *
+   * @return
+   *   Array of info, or FALSE if the record is not found.
+   */
+  function _getSystemInfo($name, $type) {
+    $raw_info = db_query("SELECT info FROM {system} WHERE name = :name AND type = :type", array(':name' => $name, ':type' => $type))->fetchField();
+    return $raw_info ? unserialize($raw_info) : FALSE;
+  }
+}
+
+/**
  * Tests for the update system functionality.
  */
 class UpdateScriptFunctionalTest extends DrupalWebTestCase {
