Index: modules/block/block.module
===================================================================
RCS file: /cvs/drupal/drupal/modules/block/block.module,v
retrieving revision 1.424
diff -u -p -r1.424 block.module
--- modules/block/block.module	17 Jun 2010 13:16:57 -0000	1.424
+++ modules/block/block.module	18 Jun 2010 20:55:47 -0000
@@ -847,12 +847,14 @@ function _block_get_cache_id($block) {
  * Implements hook_flush_caches().
  */
 function block_flush_caches() {
-  // Rehash blocks for active themes. We don't use list_themes() here,
-  // because if MAINTENANCE_MODE is defined it skips reading the database,
-  // and we can't tell which themes are active.
-  $themes = db_query("SELECT name FROM {system} WHERE type = 'theme' AND status = 1");
-  foreach ($themes as $theme) {
-    _block_rehash($theme->name);
+  // Rehash blocks for active themes after running database updates. We don't
+  // use list_themes() here, because if MAINTENANCE_MODE is defined it skips
+  // reading the database, and we can't tell which themes are active.
+  if (defined('MAINTENANCE_MODE') && MAINTENANCE_MODE == 'update') {
+    $themes = db_query("SELECT name FROM {system} WHERE type = 'theme' AND status = 1");
+    foreach ($themes as $theme) {
+      _block_rehash($theme->name);
+    }
   }
 
   return array('cache_block');
