Index: modules/system/system.install
===================================================================
RCS file: /cvs/drupal/drupal/modules/system/system.install,v
retrieving revision 1.531
diff -u -p -r1.531 system.install
--- modules/system/system.install	29 Dec 2010 03:46:39 -0000	1.531
+++ modules/system/system.install	30 Dec 2010 19:11:58 -0000
@@ -87,6 +87,20 @@ function system_requirements($phase) {
     return $requirements;
   }
 
+  // Test for eAccelerator.
+  if (function_exists('eaccelerator_info')) {
+    $info = eaccelerator_info();
+    $eaccelerator_minimum_version = '0.9.5.3';
+    if (version_compare($info['version'], $eaccelerator_minimum_version, '<')) {
+      $requirements['php_eaccelerator'] = array(
+        'title' => $t('eAccelerator'),
+        'value' => $info['version'],
+        'description' => $t('Your eAccelerator installation is too old. Drupal is only compatible with version %version or later.', array('%version' => $eaccelerator_minimum_version)),
+        'severity' => REQUIREMENT_ERROR,
+      );
+    }
+  }
+
   // Test PHP register_globals setting.
   $requirements['php_register_globals'] = array(
     'title' => $t('PHP register globals'),
