Index: includes/install.core.inc
===================================================================
RCS file: /cvs/drupal/drupal/includes/install.core.inc,v
retrieving revision 1.34
diff -u -p -r1.34 install.core.inc
--- includes/install.core.inc	24 Sep 2010 21:36:22 -0000	1.34
+++ includes/install.core.inc	30 Sep 2010 21:13:57 -0000
@@ -1772,7 +1772,7 @@ function _install_configure_form($form, 
   $form['server_settings']['clean_url'] = array(
     '#type' => 'hidden',
     '#default_value' => 0,
-    '#attributes' => array('class' => array('install')),
+    '#attributes' => array('id' => 'edit-clean-url', 'class' => array('install')),
   );
 
   $form['update_notifications'] = array(
