Index: contribute/contribute.module
===================================================================
--- contribute/contribute.module	(revision 285)
+++ contribute/contribute.module	(working copy)
@@ -247,11 +247,8 @@
   }
   $result = db_query("SELECT * FROM {contribution_funds} WHERE `name` = '%s'", $name);
   if ($fund = db_fetch_object($result)) {
-    $result = db_query("SELECT * FROM {variable} WHERE `name` LIKE 'contribute_%d_%'", $fund->fid);
-    while ($var = db_fetch_object($result)) {
-      $var_name = preg_replace("/^(contribute_". $fund->fid ."_)/", '', $var->name);
-      $fund->$var_name = @unserialize($var->value);
-    }
+    $fund = drupal_unpack($fund);
+    unset($fund->data);
     if (empty($fund->processor)) {
       foreach (module_implements('contribute_process') as $module) {
         $fund->processor = str_replace('contribute_', '', $module);
Index: contribute/contribute.install
===================================================================
--- contribute/contribute.install	(revision 285)
+++ contribute/contribute.install	(working copy)
@@ -15,7 +15,7 @@
  */
 function contribute_schema() {
   $schema['contributions'] = array(
-    'fields' => array(    
+    'fields' => array(
       'cid'             => array('type' => 'serial', 'unsigned' => TRUE, 'not null' => TRUE), // auto-incremented integer, unique contribution id
       'fid'             => array('type' => 'int', 'not null' => TRUE, 'default' => 0), // fund id related
       'scid'            => array('type' => 'varchar', 'length' => 24, 'not null' => TRUE), // source code, may have corresponding entry in {contribution_sources}
@@ -67,7 +67,8 @@
       'name'            => array('type' => 'varchar', 'length' => 24, 'not null' => TRUE), // machine-friendly name
       'title'           => array('type' => 'varchar', 'length' => 255, 'not null' => TRUE), // human-friendly title of fund
       'status'          => array('type' => 'varchar', 'length' => 255), // status, can be ACTIVE, DEFAULT, or INACTIVE
-      'created'         => array('type' => 'int', 'not null' => TRUE, 'default' => 0)
+      'created'         => array('type' => 'int', 'not null' => TRUE, 'default' => 0),
+      'data'            => array('type' => 'text', 'size' => 'medium') // variables for the fund, serialized.
     ),
     'primary key' => array('fid'),
     'unique keys' => array(
@@ -99,3 +100,22 @@
   return $updates;
 }
 
+/**
+ * Add data field to funds. Storing fund variables with variable_set() is no
+ * longer done.
+ */
+function contribute_update_6002() {
+  $updates = array();
+  $result = db_query("SELECT * FROM {contribution_funds}");
+  while ($fund = db_fetch_object($result)) {
+    $data = array();
+    $var_result = db_query("SELECT * FROM {variable} WHERE name LIKE '%s'", 'contribute_'. $fund->fid .'_%%');
+    while ($var = db_fetch_object($var_result)) {
+      variable_del($var->name);
+      $var_name = preg_replace('/^(contribute_'. $fund->fid .'_)/', '', $var->name);
+      $data[$var_name] = @unserialize($var->value);
+    }
+    $updates[] = update_sql("UPDATE {contribution_funds} SET data = '%s' WHERE fid = %d", serialize($data), $fund->fid);
+  }
+  return $updates;
+}
