Index: includes/backend.inc
===================================================================
RCS file: /cvs/drupal-contrib/contributions/modules/drush/includes/backend.inc,v
retrieving revision 1.27
diff -u -p -r1.27 backend.inc
--- includes/backend.inc	29 Jan 2010 06:14:26 -0000	1.27
+++ includes/backend.inc	4 Mar 2010 20:10:51 -0000
@@ -247,7 +247,9 @@ function _drush_proc_open($cmd, $data = 
  *   If the command was completed, this will return an associative array containing the data from drush_backend_output().
  */
 function drush_backend_invoke($command, $data = array(), $method = 'GET', $integrate = TRUE, $drush_path = NULL, $hostname = NULL, $username = NULL) {
-  $args = explode(" ", $command);
+  // Allow arguments to contain spaces if they are escaped with quotation marks.
+  $args = preg_split("/[\s,]*\\\"([^\\\"]+)\\\"[\s,]*|[\s,]*'([^']+)'[\s,]*|[\s,]+/", $command, 0, PREG_SPLIT_NO_EMPTY | PREG_SPLIT_DELIM_CAPTURE);
+  // The first component is the command.
   $command = array_shift($args);
   return drush_backend_invoke_args($command, $args, $data, $method, $integrate, $drush_path, $hostname, $username);
 }
