Index: includes/theme.inc
===================================================================
RCS file: /cvs/drupal/drupal/includes/theme.inc,v
retrieving revision 1.598
diff -u -p -r1.598 theme.inc
--- includes/theme.inc	19 May 2010 19:22:24 -0000	1.598
+++ includes/theme.inc	25 May 2010 21:07:49 -0000
@@ -2403,7 +2403,9 @@ function template_preprocess_maintenance
   $variables['language']          = $language;
   $variables['language']->dir     = $language->direction ? 'rtl' : 'ltr';
   $variables['logo']              = theme_get_setting('logo');
-  $variables['messages']          = $variables['show_messages'] ? theme('status_messages') : '';
+  if (!isset($variables['messages'])) {
+    $variables['messages']        = $variables['show_messages'] ? theme('status_messages') : '';
+  }
   $variables['main_menu']         = array();
   $variables['secondary_menu']    = array();
   $variables['site_name']         = (theme_get_setting('toggle_name') ? variable_get('site_name', 'Drupal') : '');
Index: includes/theme.maintenance.inc
===================================================================
RCS file: /cvs/drupal/drupal/includes/theme.maintenance.inc,v
retrieving revision 1.59
diff -u -p -r1.59 theme.maintenance.inc
--- includes/theme.maintenance.inc	12 May 2010 09:22:24 -0000	1.59
+++ includes/theme.maintenance.inc	25 May 2010 21:07:49 -0000
@@ -142,11 +142,12 @@ function theme_task_list($variables) {
 function theme_install_page($variables) {
   drupal_add_http_header('Content-Type', 'text/html; charset=utf-8');
 
-  // Delay setting the message variable so it can be processed below.
-  $variables['show_messages'] = FALSE;
-
   // Special handling of error messages
   $messages = drupal_set_message();
+  // Initialize variables.
+  if (!isset($variables['messages'])) {
+    $variables['messages'] = '';
+  }
   if (isset($messages['error'])) {
     $title = count($messages['error']) > 1 ? st('The following errors must be resolved before you can continue the installation process') : st('The following error must be resolved before you can continue the installation process');
     $variables['messages'] .= '<h3>' . $title . ':</h3>';
