? .cvsignore
? .project
? .settings
? test.php
Index: includes/bootstrap.inc
===================================================================
RCS file: /cvs/drupal/drupal/includes/bootstrap.inc,v
retrieving revision 1.348
diff -u -p -r1.348 bootstrap.inc
--- includes/bootstrap.inc	31 Jan 2010 18:31:46 -0000	1.348
+++ includes/bootstrap.inc	4 Feb 2010 23:09:30 -0000
@@ -2087,6 +2087,24 @@ function _registry_check_code($type, $na
     return TRUE;
   }
 
+  // Skip registry lookups for DB-related functions.
+  $connection_info = Database::getConnectionInfo();
+  $driver = $connection_info['default']['driver'];
+  if (preg_match('/_' . $driver .'$/', $name)) {
+    // Manually include database files, since the registry cannot build without
+    // these classes in memory.
+    require_once DRUPAL_ROOT . '/includes/database/query.inc';
+    require_once DRUPAL_ROOT . '/includes/database/select.inc';
+    require_once DRUPAL_ROOT . '/includes/database/schema.inc';
+    foreach (glob(DRUPAL_ROOT . '/includes/database/' . $driver . '/*.inc') as $include_file) {
+      if (basename($include_file) != 'install.inc') {
+        require_once $include_file;
+      }
+    }
+    $function = $type . '_exists';
+    return $function($name);
+  }
+
   if (!isset($lookup_cache)) {
     $lookup_cache = array();
     if ($cache = cache_get('lookup_cache', 'cache_bootstrap')) {
@@ -2124,7 +2142,22 @@ function _registry_check_code($type, $na
   // This function may get called when the default database is not active, but
   // there is no reason we'd ever want to not use the default database for
   // this query.
-  $file = Database::getConnection('default', 'default')->query("SELECT filename FROM {registry} WHERE name = :name AND type = :type", array(
+  $db = Database::getConnection('default', 'default');
+  if (!$count = $db->query('SELECT 1 FROM {registry}')->fetchField()) {
+    // The registry table is empty. Attempt to repopulate it.
+    // If Drupal hasn't been fully boostrapped, we need to include common.inc.
+    require_once DRUPAL_ROOT . '/includes/common.inc';
+    require_once DRUPAL_ROOT . '/includes/registry.inc';
+    _registry_update();
+    if (!$count = $db->query('SELECT 1 FROM {registry}')->fetchField()) {
+      exit('Registry is STILL empty!');
+    }
+  }
+
+  // This function may get called when the default database is not active, but
+  // there is no reason we'd ever want to not use the default database for
+  // this query.
+  $file = $db->query("SELECT filename FROM {registry} WHERE name = :name AND type = :type", array(
       ':name' => $name,
       ':type' => $type,
     ))
Index: includes/registry.inc
===================================================================
RCS file: /cvs/drupal/drupal/includes/registry.inc,v
retrieving revision 1.27
diff -u -p -r1.27 registry.inc
--- includes/registry.inc	28 Dec 2009 10:48:51 -0000	1.27
+++ includes/registry.inc	4 Feb 2010 23:09:30 -0000
@@ -52,6 +52,9 @@ function _registry_update() {
       }
     }
   }
+  if (!function_exists('file_scan_directory')) {
+    require_once DRUPAL_ROOT . '/includes/file.inc';
+  }
   foreach (file_scan_directory('includes', '/\.inc$/') as $filename => $file) {
     $files["$filename"] = array('module' => '', 'weight' => 0);
   }
