Index: includes/install.core.inc
===================================================================
RCS file: /cvs/drupal/drupal/includes/install.core.inc,v
retrieving revision 1.4
diff -u -p -r1.4 install.core.inc
--- includes/install.core.inc	1 Mar 2010 07:39:12 -0000	1.4
+++ includes/install.core.inc	1 Mar 2010 23:51:05 -0000
@@ -232,6 +232,11 @@ function install_begin_request(&$install
 
   require_once DRUPAL_ROOT . '/modules/system/system.install';
   require_once DRUPAL_ROOT . '/includes/common.inc';
+
+  // Ensure that PHP doesn't time out during install.
+  drupal_set_time_limit(240);
+  ignore_user_abort(TRUE);
+
   require_once DRUPAL_ROOT . '/includes/file.inc';
   require_once DRUPAL_ROOT . '/includes/install.inc';
   require_once DRUPAL_ROOT . '/' . variable_get('path_inc', 'includes/path.inc');
