Index: mongodb.module
===================================================================
RCS file: /cvs/drupal-contrib/contributions/modules/mongodb/mongodb.module,v
retrieving revision 1.5
diff -u -p -r1.5 mongodb.module
--- mongodb.module	18 Dec 2009 15:05:28 -0000	1.5
+++ mongodb.module	23 Dec 2010 02:05:00 -0000
@@ -58,3 +58,22 @@ function mongodb_test_group_finished() {
     }
   }
 }
+
+/**
+ * MongoDB version of pager_query().
+ */
+function mongodb_pager_query($collection_name, $conditions, $sort = array(), $limit = 10, $element = 0) {
+  global $pager_page_array, $pager_total, $pager_total_items;
+  $page_string = isset($_GET['page']) ? $_GET['page'] : '';
+
+  // Convert comma-separated $page to an array, used by other functions.
+  $pager_page_array = explode(',', $page_string);
+
+  // We calculate the total of pages as ceil(items / limit).
+  $cursor = mongodb_collection($collection_name)->find($conditions);
+  $pager_total_items[$element] = $cursor->count();
+  $pager_total[$element] = ceil($pager_total_items[$element] / $limit);
+  $pager_page_array[$element] = max(0, min((int)$pager_page_array[$element], ((int)$pager_total[$element]) - 1));
+  $cursor->skip($pager_page_array[$element] * $limit)->limit($limit);
+  return $cursor;
+}
Index: mongodb_watchdog/mongodb_watchdog.admin.inc
===================================================================
RCS file: /cvs/drupal-contrib/contributions/modules/mongodb/mongodb_watchdog/mongodb_watchdog.admin.inc,v
retrieving revision 1.3.2.1
diff -u -p -r1.3.2.1 mongodb_watchdog.admin.inc
--- mongodb_watchdog/mongodb_watchdog.admin.inc	8 Feb 2010 16:35:28 -0000	1.3.2.1
+++ mongodb_watchdog/mongodb_watchdog.admin.inc	23 Dec 2010 02:05:00 -0000
@@ -38,14 +38,12 @@ function mongodb_watchdog_overview() {
   );
 
   $per_page = 50;
-  $on_page = isset($_GET['page']) ? $_GET['page'] : 0;
 
-  module_load_include('module', 'mongodb');
-  $cursor = mongodb_collection(variable_get('mongodb_collectionname', 'watchdog'))
-    ->find(mongodb_watchdog_build_filter_query())
-    ->limit($per_page)
-    ->skip($on_page * $per_page)
-    ->sort(array('timestamp' => -1));
+  $cursor = mongodb_pager_query(
+    variable_get('mongodb_collectionname', 'watchdog'),
+    mongodb_watchdog_build_filter_query(),
+    array('timestamp' => -1),
+    $per_page);
 
   $build['mongodb_watchdog_filter_form']    = mongodb_watchdog_filter_form() ;
   $build['mongodb_watchdog_clear_log_form'] = mongodb_watchdog_clear_log_form();
@@ -74,7 +72,7 @@ function mongodb_watchdog_overview() {
   $attributes = array('id' => 'admin-mongodb_watchdog');
   $build['mongodb_watchdog_table'] = array(
     '#type' => 'markup',
-    '#value' => theme('table', $header, $rows, $attributes),
+    '#value' => theme('table', $header, $rows, $attributes) . theme('pager', array(), $per_page),
   );
 
   return $build;
