Index: modules/shortcut/shortcut.admin.inc
===================================================================
RCS file: /cvs/drupal/drupal/modules/shortcut/shortcut.admin.inc,v
retrieving revision 1.13
diff -u -p -r1.13 shortcut.admin.inc
--- modules/shortcut/shortcut.admin.inc	24 Apr 2010 14:49:14 -0000	1.13
+++ modules/shortcut/shortcut.admin.inc	24 May 2010 21:51:30 -0000
@@ -451,6 +451,11 @@ function _shortcut_link_form_elements($s
     '#default_value' => $shortcut_link['link_path'],
   );
 
+  // If the shortcut has a query string, make it editable as well.
+  if (!empty($shortcut_link['options']['query'])) {
+    $form['shortcut_link']['link_path']['#default_value'] .= '?' . drupal_http_build_query($shortcut_link['options']['query']);
+  }
+
   $form['#validate'][] = 'shortcut_link_edit_validate';
 
   $form['actions'] = array('#type' => 'actions');
@@ -466,6 +471,13 @@ function _shortcut_link_form_elements($s
  * Validation handler for the shortcut link add and edit forms.
  */
 function shortcut_link_edit_validate($form, &$form_state) {
+  $form_state['values']['shortcut_link']['options']['query'] = array();
+  if (strstr($form_state['values']['shortcut_link']['link_path'], '?')) {
+    list($link_path, $query_string) = explode('?', $form_state['values']['shortcut_link']['link_path']);
+    $form_state['values']['shortcut_link']['link_path'] = $link_path;
+    parse_str($query_string, $form_state['values']['shortcut_link']['options']['query']);
+  }
+
   if (!shortcut_valid_link($form_state['values']['shortcut_link']['link_path'])) {
     form_set_error('shortcut_link][link_path', t('The link must correspond to a valid path on the site.'));
   }
@@ -697,6 +709,11 @@ function shortcut_link_delete_submit($fo
  *   Returned from shortcut_set_load().
  */
 function shortcut_link_add_inline($shortcut_set) {
+  if (isset($_GET['link']) && strstr($_GET['link'], '?')) {
+    list($_GET['link'], $query_string) = explode('?', $_GET['link']);
+    $query = array();
+    parse_str($query_string, $query);
+  }
   if (isset($_REQUEST['token']) && drupal_valid_token($_REQUEST['token'], 'shortcut-add-link') && shortcut_valid_link($_GET['link'])) {
     $item = menu_get_item($_GET['link']);
     $title = ($item && $item['title']) ? $item['title'] : $_GET['name'];
@@ -704,6 +721,9 @@ function shortcut_link_add_inline($short
       'link_title' => $title,
       'link_path' => $_GET['link'],
     );
+    if (isset($query)) {
+      $link['options'] = array('query' => $query);
+    }
     shortcut_admin_add_link($link, $shortcut_set, shortcut_max_slots());
     if (shortcut_set_save($shortcut_set)) {
       drupal_set_message(t('Added a shortcut for %title.', array('%title' => $link['link_title'])));
Index: modules/shortcut/shortcut.module
===================================================================
RCS file: /cvs/drupal/drupal/modules/shortcut/shortcut.module,v
retrieving revision 1.24
diff -u -p -r1.24 shortcut.module
--- modules/shortcut/shortcut.module	6 Mar 2010 06:33:14 -0000	1.24
+++ modules/shortcut/shortcut.module	24 May 2010 21:51:30 -0000
@@ -628,6 +628,8 @@ function shortcut_preprocess_page(&$vari
   if (shortcut_set_edit_access()) {
     $link = $_GET['q'];
     $query_parameters = drupal_get_query_parameters();
+    // Allow query parameters in the shortcut link, with a few exceptions here.
+    $query_parameters = drupal_get_query_parameters(NULL, array('q', 'render', 'page'));
     if (!empty($query_parameters)) {
      $link .= '?' . drupal_http_build_query($query_parameters);
     }
@@ -641,6 +643,9 @@ function shortcut_preprocess_page(&$vari
 
     // Check if $link is already a shortcut and set $link_mode accordingly.
     foreach ($shortcut_set->links as $shortcut) {
+      if (!empty($shortcut['options']['query'])) {
+        $shortcut['link_path'] .= '?' . drupal_http_build_query($shortcut['options']['query']);
+      }
       if ($link == $shortcut['link_path']) {
         $mlid = $shortcut['mlid'];
         break;
