Index: sites/all/modules/terraeclipse/signupapi/signupapi.module
===================================================================
--- sites/all/modules/terraeclipse/signupapi/signupapi.module	(revision 2191)
+++ sites/all/modules/terraeclipse/signupapi/signupapi.module	(working copy)
@@ -435,6 +435,9 @@ function signupapi_schema_alter(&$schema) {
  */
 function signupapi_get_field_spec($field_name) {
   $schema = drupal_get_schema('signups');
+  if (!isset($schema['fields'])) {
+    return FALSE;
+  }
   $fields = $schema['fields'] + array(
     'name_optin' => array(
       'type' => 'int',
