Index: session.singlesignon.inc
===================================================================
RCS file: /cvs/drupal-contrib/contributions/modules/sso/session.singlesignon.inc,v
retrieving revision 1.1
diff -u -r1.1 session.singlesignon.inc
--- session.singlesignon.inc	30 Sep 2009 18:55:11 -0000	1.1
+++ session.singlesignon.inc	24 Feb 2010 21:15:47 -0000
@@ -218,7 +218,7 @@ function singlesignon_get_domain($url = NULL) {
     $url = $base_url;
   }
   $parsed = parse_url($url);
-  return preg_replace('/^www\./', '', $parsed['host']) . ($parsed['scheme'] == 'https' ? ':ssl' : '');
+  return preg_replace('/^www\./', '', $parsed['host']);
 }
 
 /**
