Index: memcache.inc
===================================================================
RCS file: /cvs/drupal-contrib/contributions/modules/memcache/memcache.inc,v
retrieving revision 1.15.2.8.2.21
diff -u -p -r1.15.2.8.2.21 memcache.inc
--- memcache.inc	13 Oct 2010 03:11:31 -0000	1.15.2.8.2.21
+++ memcache.inc	16 Nov 2010 05:59:49 -0000
@@ -152,6 +152,11 @@ function cache_clear_all($cid = NULL, $t
     if ($cid == '*') {
       $cid = '';
     }
+    if (empty($cid)) {
+      if ($table == 'cache_page') {
+        register_shutdown_function('cache_purge_varnish');
+      }
+    }
     if (variable_get('cache_lifetime', 0) && empty($cid)) {
       // Update the timestamp of the last global flushing of this table.  When
       // retrieving data from this table, we will compare the cache creation
@@ -278,3 +283,14 @@ function memcache_wildcards($cid, $table
 function memcache_clone($object) {
   return version_compare(phpversion(), '5.0') < 0 ? $object : clone($object);
 }
+
+/**
+ * Shutdown function which clears varnish cache on cache_clear_all().
+ */
+function cache_purge_varnish() {
+  static $purged = FALSE;
+  if (!$purged && function_exists('varnish_purge_all_pages') && variable_get('varnish_cache_clear', VARNISH_DEFAULT_CLEAR) == VARNISH_DEFAULT_CLEAR) {
+    varnish_purge_all_pages();
+    $purged = TRUE;
+  }
+}
