Index: includes/cache.inc
===================================================================
--- includes/cache.inc	(revision 1041)
+++ includes/cache.inc	(working copy)
@@ -179,5 +179,15 @@ function cache_clear_all($cid = NULL, $table = NUL
       db_query("DELETE FROM {". $table ."} WHERE cid = '%s'", $cid);
     }
   }
+  if ($table == 'cache_page') {
+    register_shutdown_function('cache_purge_varnish');
+  }
 }
 
+function cache_purge_varnish() {
+  static $purged = FALSE;
+  if (!$purged && function_exists('varnish_purge_all_pages') && variable_get('varnish_cache_clear', VARNISH_DEFAULT_CLEAR) == VARNISH_DEFAULT_CLEAR) {
+    varnish_purge_all_pages();
+    $purged = TRUE;
+  }
+}
